datafusion/functions/src/math/decimal.rs

```diff
@@ -0,0 +1,207 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::ops::Rem;
+use std::sync::Arc;
+
+use arrow::array::{ArrayRef, AsArray, PrimitiveArray};
+use arrow::datatypes::DecimalType;
+use arrow::error::ArrowError;
+use arrow_buffer::i256;
+use datafusion_common::{exec_err, Result};
+
+/// Operations required to manipulate the native representation of Arrow decimal arrays.
+pub(super) trait DecimalNative:
+    Copy + Rem<Output = Self> + PartialEq + PartialOrd
+{
+    fn zero() -> Self;
+    fn checked_add(self, other: Self) -> Option<Self>;
+    fn checked_sub(self, other: Self) -> Option<Self>;
+    fn checked_pow10(exp: u32) -> Option<Self>;
+}
+
+impl DecimalNative for i32 {
+    fn zero() -> Self {
+        0
+    }
+
+    fn checked_add(self, other: Self) -> Option<Self> {
+        self.checked_add(other)
+    }
+
+    fn checked_sub(self, other: Self) -> Option<Self> {
+        self.checked_sub(other)
+    }
+
+    fn checked_pow10(exp: u32) -> Option<Self> {
+        10_i32.checked_pow(exp)
+    }
+}
+
+impl DecimalNative for i64 {
+    fn zero() -> Self {
+        0
+    }
+
+    fn checked_add(self, other: Self) -> Option<Self> {
+        self.checked_add(other)
+    }
+
+    fn checked_sub(self, other: Self) -> Option<Self> {
+        self.checked_sub(other)
+    }
+
+    fn checked_pow10(exp: u32) -> Option<Self> {
+        10_i64.checked_pow(exp)
+    }
+}
+
+impl DecimalNative for i128 {
+    fn zero() -> Self {
+        0
+    }
+
+    fn checked_add(self, other: Self) -> Option<Self> {
+        self.checked_add(other)
+    }
+
+    fn checked_sub(self, other: Self) -> Option<Self> {
+        self.checked_sub(other)
+    }
+
+    fn checked_pow10(exp: u32) -> Option<Self> {
+        10_i128.checked_pow(exp)
+    }
+}
+
+impl DecimalNative for i256 {
+    fn zero() -> Self {
+        i256::ZERO
+    }
+
+    fn checked_add(self, other: Self) -> Option<Self> {
+        self.checked_add(other)
+    }
+
+    fn checked_sub(self, other: Self) -> Option<Self> {
+        self.checked_sub(other)
+    }
+
+    fn checked_pow10(exp: u32) -> Option<Self> {
+        i256::from_i128(10).checked_pow(exp)
+    }
+}
+
+pub(super) fn apply_decimal_op<T, F>(
+    array: &ArrayRef,
+    scale: i8,
+    fn_name: &str,
+    op: F,
+) -> Result<ArrayRef>
+where
+    T: DecimalType,
+    T::Native: DecimalNative,
+    F: Fn(T::Native, T::Native) -> std::result::Result<T::Native, ArrowError>,
+{
+    if scale <= 0 {
+        return Ok(Arc::clone(array));
+    }
+
+    let factor = decimal_scale_factor::<T>(scale, fn_name)?;
+    let decimal = array.as_primitive::<T>();
+    let data_type = array.data_type().clone();
+
+    let result: PrimitiveArray<T> = decimal
+        .try_unary(|value| op(value, factor))?
+        .with_data_type(data_type);
+
+    Ok(Arc::new(result))
+}
+
+fn decimal_scale_factor<T>(scale: i8, fn_name: &str) -> Result<T::Native>
```

Comment:
I feel we should just inline this function; doing so would make it clear that `scale` was already checked for being non-negative